src/ripple/app/paths/impl/PaySteps.cpp

```diff
@@ -370,6 +374,308 @@ toStrand (
     return {tesSUCCESS, std::move (result)};
 }
 
+
+std::pair<TER, Strand>
+toStrandV2 (
+    ReadView const& view,
+    AccountID const& src,
+    AccountID const& dst,
+    Issue const& deliver,
+    boost::optional<Issue> const& sendMaxIssue,
+    STPath const& path,
+    bool ownerPaysTransferFee,
+    beast::Journal j)
+{
+    if (isXRP(src) || isXRP(dst) ||
+        !isConsistent(deliver) || (sendMaxIssue && !isConsistent(*sendMaxIssue)))
+        return {temBAD_PATH, Strand{}};
+
+    for (auto const& p : path)
+    {
+        auto const t = p.getNodeType();
+
+        if (t > STPathElement::Type::typeAll || !t)
+            return {temBAD_PATH, Strand{}};
+
+        bool const hasAccount = t & STPathElement::Type::typeAccount;
+        bool const hasIssuer = t & STPathElement::Type::typeIssuer;
+        bool const hasCurrency = t & STPathElement::Type::typeCurrency;
+
+        if (hasAccount && (hasIssuer || hasCurrency))
+            return {temBAD_PATH, Strand{}};
+
+        if (hasIssuer && isXRP(p.getIssuerID()))
+            return {temBAD_PATH, Strand{}};
+
+        if (hasAccount && isXRP(p.getAccountID()))
+            return {temBAD_PATH, Strand{}};
+
+        if (hasCurrency && hasIssuer &&
+            isXRP(p.getCurrency()) != isXRP(p.getIssuerID()))
+            return {temBAD_PATH, Strand{}};
+    }
+
+    Issue curIssue = [&]
+    {
+        auto& currency =
+            sendMaxIssue ? sendMaxIssue->currency : deliver.currency;
+        if (isXRP (currency))
+            return xrpIssue ();
+        return Issue{currency, src};
+    }();
+
+    auto hasCurrency = [](STPathElement const pe)
+    {
+        return pe.getNodeType () & STPathElement::typeCurrency;
+    };
+
+    std::vector<STPathElement> normPath;
```

Comment:
That limit lives in `Payment.h` and is checked in the payment tx preclaim. I could do this in two ways:

1) Make that limit public and add a dependency between `PaySteps.cpp` and `Payment.h`. I don't like dependencies between lower-level routines and higher level.

2) I could set a largish limit - say 32 - and a comment that if the number of path steps are increased, that should be increased as well. That seems fragile.

While I love getting rid of memory allocations, I don't think it's a huge deal in this case. It's done once at the start of the payment. I'd probably keep it as-is.